storage/src/main/java/tech/pegasys/teku/storage/Store.java

```diff
@@ -377,6 +382,58 @@ public boolean containsCheckpointState(Checkpoint checkpoint) {
     }
   }
 
+  private Optional<BeaconState> getOrGenerateBlockState(final Bytes32 blockRoot) {
+    Optional<BeaconState> state = Optional.ofNullable(this.block_states.get(blockRoot));
+    if (state.isPresent()) {
+      return state;
+    }
+    final SignedBeaconBlock blockForState = getSignedBlock(blockRoot);
+    if (blockForState == null) {
+      // If we don't have the corresponding block, we can't possibly regenerate the state
+      return Optional.empty();
+    }
+
+    // Accumulate blocks until we find our base state to build from
+    SignedBlockAndState baseBlock = null;
+    final List<SignedBeaconBlock> blocks = new ArrayList<>();
+    SignedBeaconBlock block = blockForState;
+    while (block != null) {
+      final BeaconState blockState = this.block_states.get(block.getRoot());
+      if (blockState != null) {
+        // We found a base state
+        baseBlock = new SignedBlockAndState(block, blockState);
+        break;
+      }
+      blocks.add(block);
+      block = getSignedBlock(block.getParent_root());
+    }
+
+    if (baseBlock == null) {
+      // If we haven't found a base state yet, we must have walked back to the latest finalized
+      // block, check here for the base state
+      final BeaconState latestFinalizedState = checkpoint_states.get(finalized_checkpoint);
+      final SignedBeaconBlock lastBlock = blocks.get(blocks.size() - 1);
+      if (!lastBlock.getMessage().getState_root().equals(latestFinalizedState.hash_tree_root())) {
+        // Something went wrong - we can't find the base state
+        throw new IllegalStateException("Unable to find ancestor state for block " + blockRoot);
+      }
+      baseBlock = new SignedBlockAndState(lastBlock, latestFinalizedState);
+    }
+
+    // Regenerate state
+    final Map<Bytes32, BeaconState> regeneratedStates =
+        StateTransition.produceStatesForBlocks(baseBlock.getRoot(), baseBlock.getState(), blocks);
+
+    // Save regenerated state
+    final BeaconState regeneratedState = regeneratedStates.get(blockRoot);
+    if (regeneratedState == null) {
+      throw new IllegalStateException("Unable to generate state for block " + blockRoot);
+    }
+    this.block_states.put(blockRoot, regeneratedState);
```

Comment:
Yeah I think storing only the requested state is the best approach.  We might be doing this regeneration as part of a REST API request and wouldn't want to replace all the latest states with old ones because someone queried something just before the state cache was available.